src/analyz/mod.rs

```diff
@@ -777,15 +777,62 @@ fn emit_trait<'tcx>(
     ti: TraitInst<'tcx>,
 ) -> io::Result<()> {
     let tcx = ms.tcx;
-    let methods = if let Some(tref) = ti.concrete_trait_ref(tcx) {
+    let trait_ref = ti.concrete_trait_ref(tcx);
+
+    let methods = if let Some(tref) = trait_ref {
         tcx.vtable_entries(tref)
     } else {
         &[]
     };
-    let mut items = Vec::with_capacity(methods.len());
+
+    // Reserve space for all methods plus the additional drop function.  Usually
+    // `methods` will contain a few non-`VtblEntry::Method` entries, giving us
+    // extra slots to work with, but we don't rely on that.
+    let mut items = Vec::with_capacity(methods.len() + 1);
+
+    // To get around an issue in which rustc seems not to generate proper drop
+    // glue for trait objects, we always include a drop method in emitted traits
+    // (and vtables - see `build_vtable_items` below), and we always put it at
+    // the beginning of the method list so that `crucible-mir` can find and call
+    // it manually. We adjust `InstanceKind::Virtual` indices to account for
+    // this extra method in `ty_json::adjust_method_index`.
+    if let Some(tref) = trait_ref {
+        let dyn_ty = tref.self_ty();
+        let drop_inst = ty::Instance::resolve_drop_in_place(tcx, dyn_ty);
+        let drop_def_id = drop_inst.def.def_id();
+        let drop_args = drop_inst.args;
+        let drop_sig = tcx.instantiate_and_normalize_erasing_regions(
+            drop_args,
+            ty::TypingEnv::fully_monomorphized(),
+            tcx.fn_sig(drop_def_id),
+        );
+        let drop_sig = tcx.instantiate_bound_regions_with_erased(drop_sig);
+        items.push(json!({
+            "kind": "Method",
+            "item_id": drop_def_id.to_json(ms),
+            "signature": drop_sig.to_json(ms),
+        }));
+    } else {
+        // I believe the only way this can happen is if this trait instance was
+        // derived from predicates (`TraitInst::from_dynamic_predicates`) with
+        // no principal trait, only marker traits. For example, `dyn Debug +
+        // Send + Sync`'s principal trait would be `Debug`, but `dyn Send +
+        // Sync` would have no principal trait, because `Send` and `Sync` are
+        // marker traits.
+        eprintln!("warning: no trait ref for {:?}?", ti);
```

Comment:
Should note explicitly that we decided not to support this for now, since we don't expect to see it in real code (because it doesn't seem useful)

Reply:
Done.